[CA-1464] logging delete workspace failures to help debug CA-1206

andy7i opened the pull request:
CA-1464

Changelog:
* Adds logging messages in deleteWorkspace so it is easier for us to track down the root cause of the failed deleteWorkspace calls. 
* makes deleting in Workspace Manager fail if it's not the expected 404. (was this a bug?)
* removes the mixing of try (sync) and future (async) 

---

error message examples:

```
// before:
rawls [ERROR] [08/05/2021 16:31:33.696] [scala-execution-context-global-76] [akka.actor.ActorSystemImpl(rawls)] HttpMethod(DELETE) http://rawls-sbt:8080/api/workspaces/aj-ca-1206-6/aj-test-ws-2020-copy7: 500 Internal Server Error entity: {"causes":[],"message":"aj-test-failures","source":"rawls","stackTrace":[]}
```

```
// after:
rawls [ERROR] [16:31:33.560] [scala-execution-context-global-73] o.b.d.r.workspace.WorkspaceService - Unexpected failure deleting workspace (while deleting workflowCollection in Sam) for workspace `aj-ca-1206-6/aj-test-ws-2020-copy7`
rawls org.broadinstitute.dsde.rawls.RawlsExceptionWithErrorReport: ErrorReport(rawls,aj-test-failures,None,List(),List(),None)
rawls 	at org.broadinstitute.dsde.rawls.dataaccess.HttpSamDAO.$anonfun$deleteResource$1(HttpSamDAO.scala:159)
rawls 	at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:659)
rawls 	at scala.util.Success.$anonfun$map$1(Try.scala:255)
rawls 	at scala.util.Success.map(Try.scala:213)
rawls 	at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
rawls 	at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
rawls 	at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
rawls 	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
rawls 	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1426)
rawls 	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
rawls 	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
rawls 	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
rawls 	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
rawls 	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
rawls [WARN] [16:31:33.648] [scala-execution-context-global-73] io.sentry.dsn.Dsn - *** Couldn't find a suitable DSN, Sentry operations will do nothing! See documentation: https://docs.sentry.io/clients/java/ ***
rawls [ERROR] [08/05/2021 16:31:33.696] [scala-execution-context-global-76] [akka.actor.ActorSystemImpl(rawls)] HttpMethod(DELETE) http://rawls-sbt:8080/api/workspaces/aj-ca-1206-6/aj-test-ws-2020-copy7: 500 Internal Server Error entity: {"causes":[],"message":"aj-test-failures","source":"rawls","stackTrace":[]}
```


---

I decided not to refactor deleteWorkspace too much right now, since we should probably figure out the root cause first before we make a lot of changes.

---

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

andy7i committed:
Try runs sync -- not good practice to mix Try with Futures

andy7i committed:
seems to me like this should only succeed if it encounters the expected error, and not all errors.

andy7i committed:
error

andy7i committed:
add error logging

andy7i committed:
add error logging

```
// before:
rawls [ERROR] [08/05/2021 16:31:33.696] [scala-execution-context-global-76] [akka.actor.ActorSystemImpl(rawls)] HttpMethod(DELETE) http://rawls-sbt:8080/api/workspaces/aj-ca-1206-6/aj-test-ws-2020-copy7: 500 Internal Server Error entity: {"causes":[],"message":"aj-test-failures","source":"rawls","stackTrace":[]}
```

```
// after:
rawls [ERROR] [16:31:33.560] [scala-execution-context-global-73] o.b.d.r.workspace.WorkspaceService - Unexpected failure deleting workspace (while deleting workflowCollection in Sam) for workspace `aj-ca-1206-6/aj-test-ws-2020-copy7`
rawls org.broadinstitute.dsde.rawls.RawlsExceptionWithErrorReport: ErrorReport(rawls,aj-test-failures,None,List(),List(),None)
rawls 	at org.broadinstitute.dsde.rawls.dataaccess.HttpSamDAO.$anonfun$deleteResource$1(HttpSamDAO.scala:159)
rawls 	at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:659)
rawls 	at scala.util.Success.$anonfun$map$1(Try.scala:255)
rawls 	at scala.util.Success.map(Try.scala:213)
rawls 	at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
rawls 	at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
rawls 	at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
rawls 	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
rawls 	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1426)
rawls 	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
rawls 	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
rawls 	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
rawls 	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
rawls 	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
rawls [WARN] [16:31:33.648] [scala-execution-context-global-73] io.sentry.dsn.Dsn - *** Couldn't find a suitable DSN, Sentry operations will do nothing! See documentation: https://docs.sentry.io/clients/java/ ***
rawls [ERROR] [08/05/2021 16:31:33.696] [scala-execution-context-global-76] [akka.actor.ActorSystemImpl(rawls)] HttpMethod(DELETE) http://rawls-sbt:8080/api/workspaces/aj-ca-1206-6/aj-test-ws-2020-copy7: 500 Internal Server Error entity: {"causes":[],"message":"aj-test-failures","source":"rawls","stackTrace":[]}
```

s-rubenstein commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -348,26 +348,52 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
 
-      workflowsToAbort <- deletionFuture
+      workflowsToAbort <- deletionFuture recoverWith {
+        case t:Throwable => { // logging added as part of investigation into https://broadworkbench.atlassian.net/browse/CA-1206
```
Nit: I don't think you need these comments. Regardless of why the logs were added, they seem like good things to have around, and I would want to keep them around even after 1206 merges.

jacmrob replied:
+1

s-rubenstein reviewed (approved):
This looks good

s-rubenstein commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -348,26 +348,52 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
+            Future.successful()
           }
-          Success(())
+          throw e
```
Why do we want to throw here where we didn't before? Is it just for our records?

andy7i committed:
oops

andy7i commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -348,26 +348,52 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
       // Delete workspace manager record (which will only exist if there had ever been a TDR snapshot in the WS)
-      _ = Try(workspaceManagerDAO.deleteWorkspace(workspaceContext.workspaceIdAsUUID, userInfo.accessToken)).recoverWith {
+      _ <- Future(workspaceManagerDAO.deleteWorkspace(workspaceContext.workspaceIdAsUUID, userInfo.accessToken)).recoverWith {
         //this will only ever succeed if a TDR snapshot had been created in the WS, so we gracefully handle all exceptions here
         case e: ApiException => {
           if(e.getCode != StatusCodes.NotFound.intValue) {
-            logger.warn(s"Unexpected failure deleting workspace in Workspace Manager. Received ${e.getCode}: [${e.getResponseBody}]")
+            logger.warn(s"Unexpected failure deleting workspace (while deleting in Workspace Manager) for workspace `${workspaceName}. Received ${e.getCode}: [${e.getResponseBody}]")
+            Future.successful()
           }
-          Success(())
+          else throw e
```
Looking for a review from AS on this, to make sure this new behavior (throwing exception on non-404s) looks reasonable.

davidangb replied:
* is there a reason to move this from a `Try` to a `Future`? Either works, and if you want to use `Future` for consistency, I understand … is there some other reason?
* the logic as you have written it is incorrect; you are swallowing all non-404 errors and rethrowing all 404s. This is backwards - we should swallow 404s. I think this is why you have so many test failures.
* you're correct that you're making this part of the code stricter. As originally written, we swallowed ALL errors but only logged the non-404s; now we will throw on non-404s (after fixing the previous bullet point). This feels like a good change, now that workspace manager has been live and stable for a while.

jacmrob commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -348,26 +348,52 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
+            Future.successful()
           }
-          Success(())
+          else throw e
```
why throw the error here?

andy7i replied:
I think this may be a bug. The old behavior would return success on ANY exception. I think the intention was probably to return success ONLY on 404. We were hiding any exceptions that weren't 404s.

andy7i replied:
updated this logic based on David An's recommendation

jacmrob reviewed (commented):
mostly LG,  just one question

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -348,26 +348,52 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
+      _ <- Future(workspaceManagerDAO.deleteWorkspace(workspaceContext.workspaceIdAsUUID, userInfo.accessToken)).recoverWith {
         //this will only ever succeed if a TDR snapshot had been created in the WS, so we gracefully handle all exceptions here
         case e: ApiException => {
           if(e.getCode != StatusCodes.NotFound.intValue) {
-            logger.warn(s"Unexpected failure deleting workspace in Workspace Manager. Received ${e.getCode}: [${e.getResponseBody}]")
+            logger.warn(s"Unexpected failure deleting workspace (while deleting in Workspace Manager) for workspace `${workspaceName}. Received ${e.getCode}: [${e.getResponseBody}]")
+            Future.successful()
           }
-          Success(())
+          else throw e
```
how about this for the `recoverWith` handler?
```
        //this will only ever succeed if a TDR snapshot had been created in the WS, so we gracefully handle 404 exceptions here
        case e: ApiException if e.getCode == StatusCodes.NotFound.intValue => {
          Future.successful()
        }
        case t:Throwable => {
          logger.warn(s"Unexpected failure deleting workspace (while deleting in Workspace Manager) for workspace `${workspaceName}`", t)
          throw t
        }
```

andy7i replied:
love it, thank you!

> the logic as you have written it is incorrect; you are swallowing all non-404 errors and rethrowing all 404s. This is backwards - we should swallow 404s. I think this is why you have so many test failures.

I missed the `!=` in the original code :/

> is there a reason to move this from a Try to a Future? Either works, and if you want to use Future for consistency, I understand … is there some other reason?

Rob and I looked at this code together and explained that we shouldn't be mixing Trys (which are synchronous) and Futures (which are async) if possible.

andy7i committed:
minor text fixes

andy7i committed:
PR feedback - suggested change from David An

davidangb reviewed (approved):
thank you!

andy7i committed:
73 tests failed across orch / rawls tests. Looking at the logs, I see messages like the following, which points to us catching the wrong error code:

```
[WARN] [21:21:34.021] [scala-execution-context-global-87] o.b.d.r.workspace.WorkspaceService - Unexpected failure deleting workspace (while deleting in Workspace Manager) for workspace `gpalloc-qa-master-thvzmwl/AuthDomains b8fce396-d83e-4ff3-a61b-259af80ee707`
                                                                                bio.terra.workspace.client.ApiException: <!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
                                                                                <html><head><script src="https://us.jsagent.tcell.insight.rapid7.com/tcellagent.min.js" tcellappid="FCNonprod-NaVu9" tcellapikey="AQQBBAFLGLOxL7VE9IF9ESlLvCxD5Ykr_7xkQKq_rgn_P58IWjOhOzIh6p3aI4pTWaprlUw" tcellbaseurl="https://us.agent.tcell.insight.rapid7.com/api/v1"></script>
                                                                                <title>401 Unauthorized</title>

```

andy7i commented:
~pushed up changes and they appear on the [branch](https://github.com/broadinstitute/rawls/tree/aj--logging-delete-workspace-failures--ca-1464) but not in the PR probably due to a github outage https://www.githubstatus.com/incidents/rmfrw9dfbtbp~
the missing commit appears above this comment now

andy7i committed:
minor text fixes (pushed up changes and they appear on the branch but not in the PR)

davidangb reviewed (changes requested):
changing 404 to 401 on the delete-workspace call to WSM is not correct. WSM returns a 404 if the workspace is not found.

In the example you used, you reference a 401 code that includes HTML in its response. HTML responses are only returned by Apache. This is a sign that something else was wrong in the fiab.

davidangb commented:
wait … WSM isn't even present in FiaBs. So the FiaB tests are calling … some instance of WSM that isn't meant to be called by tests. I'm not surprised there's lots of failures.

andy7i committed:
Revert "minor text fixes (pushed up changes and they appear on the branch but not in the PR)"

This reverts commit 3d09a9a4

andy7i committed:
Revert "73 tests failed across orch / rawls tests. Looking at the logs, I see messages like the following, which points to us catching the wrong error code:"

This reverts commit d396fce1

andy7i committed:
Revert "PR feedback - suggested change from David An"

This reverts commit dec15415

andy7i committed:
Revert "oops"

This reverts commit d20c86a9

andy7i committed:
Revert "seems to me like this should only succeed if it encounters the expected error, and not all errors."

This reverts commit e3906fe2

andy7i commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -348,26 +348,51 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
-      _ = Try(workspaceManagerDAO.deleteWorkspace(workspaceContext.workspaceIdAsUUID, userInfo.accessToken)).recoverWith {
+      _ <- Future(workspaceManagerDAO.deleteWorkspace(workspaceContext.workspaceIdAsUUID, userInfo.accessToken)).recoverWith {
         //this will only ever succeed if a TDR snapshot had been created in the WS, so we gracefully handle all exceptions here
         case e: ApiException => {
           if(e.getCode != StatusCodes.NotFound.intValue) {
-            logger.warn(s"Unexpected failure deleting workspace in Workspace Manager. Received ${e.getCode}: [${e.getResponseBody}]")
+            logger.warn(s"Unexpected failure deleting workspace (while deleting in Workspace Manager) for workspace `${workspaceName}. Received ${e.getCode}: [${e.getResponseBody}]")
           }
-          Success(())
+          Future.successful()
```
@davidangb reverted the behavior here back so it logs exceptions but does not throw exceptions.

davidangb reviewed (approved):
code looks good, :+1: assuming tests pass. Thanks for sticking with this!
